RetroAchievements/RAWeb
# feat: set up basic Inertia l10n/i18n, use lang strings on `/forums/recent-posts`

This PR adds basic internationalization support to the Inertia.js+React code, all while having the capability to leverage Laravel's localization (l10n) and internationalization (i18n) JSON files in the _lang_ directory.

This is by no means a comprehensive i18n/l10n solution. The objective is to play nicely with Laravel's built-in localization system and have the ability to move forward without hardcoding strings in the React UI.

## How to test this PR

Open _lang/en_US.json_. Modify some of the values, for example:
```
    "Recent Posts": "Recent Posts123"
```
Navigate to `/forums/recent-posts`, or if you're already on it, refresh the page (Vite will not hot reload this change). You should see the text on the page has changed.

You can also try changing your locale. If you open the already-existing _pt_BR.json_ and add the following content:
```
    ":count posts in the last 24 hours": ":count posts nas últimas 24 horas",
    ":count posts in the last 7 days": ":count posts nos últimos 7 dias",
    "Additional Posts": "Posts adicionais",
    "Forum Index": "Índice do Fórum",
    "in": "em",
    "Last Post By": "Última postagem por",
    "Message": "Mensagem",
    "Next :count": "Próximo :count",
    "Previous :count": "Anterior :count",
    "Recent Posts": "Posts Recentes"
```
You can then open _resources/js/common/components/AppProviders/AppProviders.tsx_ and change the `locale` prop on line 11 to "pt_BR". You will then see:
![Screenshot 2024-08-26 at 6 21 12 PM](https://github.com/user-attachments/assets/603cd47d-4ff6-4a75-99eb-846c1d7f1198)


## How it works
The entire app is wrapped in a `LaravelReactI18nProvider` component like:
```tsx
<LaravelReactI18nProvider>
  <div>{/* THE WHOLE APP IS HERE */}</div>
</LaravelReactI18nProvider>
```

This component injects strings from the user's current locale into the component tree and works with server-side rendering. These strings can be placed in the DOM using a single React hook: `useLaravelReactI18n()`.

```tsx
const { t, tChoice } = useLaravelReactI18n();

t("key"); // outputs the value of "key" from en_US.json

t('Welcome, :name!', { name: 'Francisco' }); // "Welcome, Francisco!"
t(':count apples', { count: 2 }); // "2 apples"

tChoice('There is one apple|There are many apples', 1); // "There is one apple"
tChoice('There is one apple|There are many apples', 4); // "There are many apples"

tChoice('{0} There are none|[1,19] There are some|[20,*] There are many', 0); // "There are none"
tChoice('{0} There are none|[1,19] There are some|[20,*] There are many', 4); // "There are some"
tChoice('{0} There are none|[1,19] There are some|[20,*] There are many', 20); // "There are many"

return (
  <p>
    {t("There are :count apples", { count: 2 })}
  </p>
);
```
